src/Mapper/TreeMapper.php

```diff
@@ -7,11 +7,18 @@
 interface TreeMapper
 {
     /**
-     * @template T of object
+     * @psalm-template TObject of object
+     * @psalm-template TypeDefinition of string|class-string<TObject>
      *
-     * @param string|class-string<T> $signature
+     * @param TypeDefinition $signature
```

Comment:
```suggestion
     * @param string|class-string<TObject> $signature
```
PhpStorm doesn't recognize the Psalm template and will not provide auto-completion if used. I'd prefer to keep this signature (unless you did it for a particular reason?). Also, the `@psalm-template TypeDefinition of string|class-string<TObject>` above should be removed.

Reply:
The point is giving PHPStan/Psalm more precise types, not to help PHPStorm.

If PHPStorm support is wished, I can provide a `.phpstorm.meta.php`, which does the auto-completion part: PHPStorm will eventually catch up with PHPStan/Psalm on this stuff.

Reply:
I'll see if I can simplify the signature first, BTW :+1:

Reply:
> The point is giving PHPStan/Psalm more precise types, not to help PHPStorm.

Sure, I got that, it's just that I believe we can have both PHPStan/Psalm support **and** PhpStorm autocompletion (very convenient for daily usage IMO).

The signature `string|class-string<TObject>` is understood by PhpStorm, and it looks like it would be ok for PHPStan/Psalm as well. Tell me if I'm wrong :D

Reply:
I just need to see if I can use the conditional here :+1:

Reply:
0bf2a13b8feac683276d8930da428f93c10246f8 simplifies this: solved